lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -1005,6 +1034,20 @@ void PrimTupleIndexOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   });
 }
 
+//===----------------------------------------------------------------------===//
+// PrimUninitializedOp
+//===----------------------------------------------------------------------===//
+
+void PrimUninitializedOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                                    MLIRContext *context) {
+  patterns.add(+[](PrimUninitializedOp op, PatternRewriter &rewriter) {
+    if (! op.use_empty())
```

Comment:
We shouldn't need to add such patterns because the canonicalizer should remove used ops without side effects automatically. I think the reason this is not removed is that the op might have side affects. 

Is this op on some dead code path?

Reply:
For the e2e example, without this a `"prim.Uninitialized"` is present unconditionally but the result is unused. I cannot track down a side-effect that would require it to remain present. Any hints as to where it make be treated as having a (pseudo) side-effect that prevents it being otherwise removed?

Reply:
`side-effect` doesn't mean there must be uses. It could just be an op that can through exception or can allocate memory. If an op doesn't have side effect, the op definition need to contain the `NoSideEffect` trait like this one https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/include/torch-mlir/Dialect/Torch/IR/TorchOps.td#L798.  Currently, we don't support generating that trait for auto generated ops ods yet. 

However, putting all these aside,  according to https://discuss.pytorch.org/t/keyerror-prim-uninitialized/64860 `prim.Uninitialized` means the compiler believe the code can never reach this statement. It's very weird to see the `prim.Uninitialzied` on an unconditional path.

Reply:
Oh, actually, it didn't say it's on dead code path and but the value is not used. Hmm, in this case we could just delete the op unconditionally for now.

Reply:
I'd seen that discussion and considered trying to remove it earlier than the point here. However in the incoming IR the result SSA is referenced, but after some folding and simplifications the unused `prim.Uninitialized` remains. This was the first point I could identify that it was straightforward to remove it. I'm happy to dig further, but would appreciate some pointers if you have them.

Reply:
I see. yea, we can leave it as it is then. btw, did you use `clang-format`? seems there is a space after `!`

Reply:
I'd run it manually but it looks like I forgot to do so after a change. Do you use https://pre-commit.com/ or similar to avoid forgetting?